common/utils/remoteServiceConnector.go

```diff
@@ -0,0 +1,109 @@
+package utils
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/auto"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+)
+
+// RemoteServiceConnector is a convenient struct to help implementing GetClientDataDeserializer().
+// Here is an example of usage when you want to serialize a Connection and a string.
+//
+//	 type ClientData struct {
+//	     MyString   string
+//	     Connection utils.Connection
+//	 }
+//
+//	 type MyRemoteService {
+//	     // This implement GetClientDataDeserializer() func(auto.UpResult) (*ClientData, error)
+//	     *RemoteServiceConnector[ClientData]
+//	 }
+//
+//		remoteServiceConnector := utils.NewRemoteServiceConnector(ctx, ClientData{})
+//		remoteServiceConnector.Register(stackKey, "Connection", connection)
+//		remoteServiceConnector.Register("test", "MyString", pulumi.String("Hello"))
+//
+//		return &MyRemoteService{
+//		    MyRemoteService: remoteServiceConnector
+//		}
+type RemoteServiceConnector[T any] struct {
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**❓ question**
There is no way in golang to limit T to structs made of `pulumi.Input`like [type guards in TS](https://www.typescriptlang.org/docs/handbook/2/narrowing.html), right ?